man/man3/Pkgs.3.md
# Pkgs

[NAME](#name)\
[SYNOPSIS](#synopsis)\
[SUBROUTINES](#subroutines)\
[get_available_updates](#get_available_updates)\
[get_inst_names](#get_inst_names)\
[get_installed_cpans](#get_installed_cpans)\
[get_installed_packages](#get_installed_packages)\
[get_local_outdated_versions](#get_local_outdated_versions)\
[get_removed_builds](#get_removed_builds)\
[EXIT CODES](#exit-codes)\
[SEE ALSO](#see-also)\
[AUTHORS](#authors)\
[LICENSE](#license)

------------------------------------------------------------------------

## NAME

SBO::Lib::Pkgs − Routines for interacting with the Slackware package
database.

## SYNOPSIS

    use SBO::Lib::Pkgs qw/ get_installed_packages /;\
    my \@installed_sbos = get_installed_packages(\'SBO\');

## SUBROUTINES

### get_available_updates

    my \@updates = \@{ get_available_updates() };

get_available_updates() compares version and build number information
for packages installed with the \_SBo tag with the local repository. It
returns an array reference to an array of hash references specifying
package names, installed versions and available versions.

### get_inst_names

    my \@names = get_inst_names(get_available_updates());

get_inst_names() returns a list of package names from an array
reference, such as one returned by get_available_updates().

### get_installed_cpans

    my \@cpans = \@{ get_installed_cpans() };

get_installed_cpans() returns an array reference to a list of Perl
modules installed from the CPAN. This is used in [sboinstall(1)](sboinstall.1.md) to
prevent conflicting installations from the CPAN and SlackBuilds.

### get_installed_packages

    my \@packages = \@{ get_installed_packages($type) };

get_installed_packages() returns an array reference to a list of
installed packages matching the specified \$type. The available types
are \"STD\" for non-SBo packages, \"SBO\" for in-tree \_SBo packages,
\"DIRTY\" for out-of-tree \_SBo packages and \"ALL\" for all.

The returned array reference holds a list of hash references
representing the names, versions and full installed package names of the
returned packages.

### get_local_outdated_versions

    my \@outdated = get_local_outdated_versions($filter);

get_local_outdated_versions() checks installed SBo packages from
\"LOCAL_OVERRIDES\". It returns an array with information about those
that have version or build numbers differing from the local repository
or the SlackBuild in \"LOCAL_OVERRIDES\".

Build number differences with the SBo repository are returned, but are
currently unused. This subroutine is used only by [sbocheck(1)](sbocheck.1.md).

### get_removed_builds

    my \@removed = get_removed_builds();

get_removed_builds() returns an array of SlackBuild names and versions
of all out-of-tree installed packages marked \"\_SBo\".

## EXIT CODES

Pkgs.pm subroutines can return the following exit code:

    \_ERR_SCRIPT 2 script or module bug

## SEE ALSO

[SBO::Lib(3)](Lib.3.md), [SBO::Lib::Build(3)](Build.3.md), [SBO::Lib::Download(3)](Download.3.md),
[SBO::Lib::Info(3)](Info.3.md), [SBO::Lib::Pkgs(3)](Pkgs.3.md), [SBO::Lib::Readme(3)](Readme.3.md),
[SBO::Lib::Repo(3)](Repo.3.md), [SBO::Lib::Tree(3)](Tree.3.md), [SBO::Lib::Util(3)](Util.3.md)

## AUTHORS

SBO::Lib was originally written by Jacob Pipkin \<j (at) dawnrazor (dot)
net\> with contributions from Luke Williams \<xocel (at) iquidus (dot)
org\> and Andreas Guldstrand \<andreas (dot) guldstrand (at) gmail (dot)
com\>.

SBO::Lib is maintained by K. Eugene Carlson \<kvngncrlsn (at) gmail
(dot) com\>.

## LICENSE

The sbotools are licensed under the MIT License.

Copyright (C) 2012−2017, Jacob Pipkin, Luke Williams, Andreas
Guldstrand.

Copyright (C) 2024−2025, K. Eugene Carlson.

------------------------------------------------------------------------


ChangeLog.md
# ChangeLog
* 1.0 - 2024-11-01
  * Change the git branch to match the Slackware version, or to a user-configured branch
  * Change the default URLs to support Slackware 15.0 and use git repositories
  * Use rsync default mirrors with RSYNC_DEFAULT (except for -current)
  * Detect build number changes and optionally upgrade
  * Detect and report out-of-tree _SBo SlackBuilds with sbocheck
  * Save any build options when running a SlackBuild
  * Install bash completions; zsh completions have been updated
  * Install a default configuration file; tweak sboconfig to play somewhat nicely with it
  * Fix making -compat32 packages for non-i486 SlackBuilds
  * Fix building -compat32 packages from templates

## Historical Changes
* 2.7 - 2019-04-28
  * Actually fix the sbofind -e bug #71
      
* 2.6 - 2019-04-27
  * Compatibility with new perl versions where you need to escape { in regexp
    #75 #77 #78
  * Add a --tries 5 option when downloading from sbosrcarch, which is a
    saner limit than the default of 20. #79
  * Change sboclean options --clean-dist and --clean-work to shorter forms #52
  * Add limited -current support using ponce's SBo repo for -current #73
  * Fix bug with sboinstall --reinstall -r #72
  * Fix bug with sbofind -e #71

* 2.5 - 2018-02-14
  * Document download behaviour #66
  * Remake sbosnap and sboremove to have OO semantics
  * Strip -compat32 from slackbuild names when looking them up #65
  * Optimise searching in sbofind

* 2.4 - 2017-05-18
  * Rewrite sboremove from the ground up so it relies less on global state
  * Fix for parsing README with useradd/groupadd commands which span lines #57
  * Add --reinstall option to sboinstall #58
  * Exit with error when sbosnap fails to sync with a repo #61
  * Add version information to sbofind output #60

* 2.3 - 2017-01-21
  * Bugfix for parsing .info files with \ among the separators #55

* 2.2 - 2017-01-17
  * Bugfix for parsing .info files with trailing whitespace after a value #54

* 2.1 - 2017-01-14
  * Internals:
    - Adding internal documentation
    - Extract code to submodules for easier separation of concerns
  * New features:
    - Support for templates for installing things with specified options #38
    - Display other README files if the slackbuild comes with them #49
  * Bugfixes
    - sboinstall/sboremove disagreeing about a package being installed #44
    - sbocheck and sboupgrade misinterpreting version strings #45
    - parsing .info files without leading space on second line #46
    - local git repo gets partially chowned to root #47
    - stop excluding .tar.gz files when rsyncing #53

* 2.0 - 2016-07-02
  * Major new features
    * LOCAL_OVERRIDES setting

      Allows to keep a directory with local slackbuild dirs that will override
      whatever is found in the regular repository. #8 #13 #14 #15 #19 #20
    * SLACKWARE_VERSION setting

      Allows to specify the slackware version to sync from SBo. Previously only
      the version in your /etc/slackware-version was used for this, and if that
      had gotten updated in -current, you'd have needed to wait both for a new
      version of sbotools, as well as SBo to get the new repository online
      before sbotools would work for you again. #19
    * REPO setting

      This will override the SLACKWARE_VERSION setting. It's used to specify an
      absolute URL for the SBo repository you want to sync with. #6 #19 #27
    * Use sbosrcarch source archive if download fails #7 #19 #24
    * sboupgrade --all option to upgrade everything listed by sbocheck. #9 #19
    * Travis CI integration

      Every push will now cause the test-suite to be run. #18
    * Hundreds of new unit-tests. #18 #19 #23 #24 #25 #27 #28 #31 #32 #33 #35 #41 #43
    * sbofind will now also use tags if they're available #37
  * Minor/bugfixes/documentation fixes
    * Use system perl when running and installing sbotools.
    * sbocheck output changed. #10 #13 #20
    * Better debug messages on errors. #16
    * manpage fixes. #17
    * sboupgrade handles dependencies better. #12 #28
    * Update bundled Sort::Versions to 1.62.
    * sboinstall/upgrade/sbocheck: small bugfixes. #21 #35 #41 #43
    * sbosnap: display download progress, update git trees better. #26 #27

* 1.9 - 2015-11-27
  * Make it compatible with perls newer than 5.18
  * Lots of code cleanup
  * Rewrite build-queue code. #2
  * Fix issue when TMP is set. #4
  * Fix various bugs related to cleanup code
  * Change location of website
  * Fix downloading of multiple sources in newer slackbuilds. #5



SBO-Lib/lib/SBO/Lib/Pkgs.pm
package SBO::Lib::Pkgs;

use 5.016;
use strict;
use warnings;

our $VERSION = '3.4.1';

use SBO::Lib::Util qw/ %config build_cmp script_error open_read version_cmp /;
use SBO::Lib::Tree qw/ get_sbo_location get_sbo_locations is_local /;
use SBO::Lib::Info qw/ get_orig_build_number get_orig_version get_sbo_build_number get_sbo_version /;

use Exporter 'import';

our @EXPORT_OK = qw{
  get_available_updates
  get_inst_names
  get_installed_cpans
  get_installed_packages
  get_local_outdated_versions
  get_removed_builds
};

our %EXPORT_TAGS = (
  all => \@EXPORT_OK,
);

=pod

=encoding UTF-8

=head1 NAME

SBO::Lib::Pkgs - Routines for interacting with the Slackware package database.

=head1 SYNOPSIS

  use SBO::Lib::Pkgs qw/ get_installed_packages /;

  my @installed_sbos = get_installed_packages('SBO');

=head1 SUBROUTINES

=cut

my $pkg_db = '/var/log/packages';

=head2 get_available_updates

  my @updates = @{ get_available_updates() };

C<get_available_updates()> compares version and build number information for
packages installed with the _SBo tag with the local repository. It returns
an array reference to an array of hash references specifying package names,
installed versions and available versions.

=cut

# for each installed sbo, find out whether or not the version or build number in
# the tree is newer, and compile an array of hashes containing those which are.
# Takes BUILD for build number only, VERS for version only and BOTH for both
sub get_available_updates {
    script_error('get_available_updates requires an argument.') unless @_ == 1;

    my $filter = shift;
    my @updates;
    my $pkg_list = get_installed_packages('SBO');

    for my $pkg (@$pkg_list) {
        my $location = get_sbo_location($pkg->{name});
        next unless $location;

        my $version = get_sbo_version($location);
        next unless $version;
        my $bump = get_sbo_build_number($location);
        next unless $bump;
        my $version_needed;
        my $build_needed;
        if ($config{STRICT_UPGRADES} eq 'TRUE' and not is_local($pkg->{name})) {
            $version_needed = version_cmp($version, $pkg->{version}) > 0;
            $build_needed = build_cmp($bump, $pkg->{numbuild}, $version, $pkg->{version}) > 0;
        } else {
            $version_needed = version_cmp($version, $pkg->{version}) != 0;
            $build_needed = build_cmp($bump, $pkg->{numbuild}, $version, $pkg->{version}) != 0;
        }
        if ($filter eq 'VERS') {
            if ($version_needed) {
                push @updates, { name => $pkg->{name}, installed => $pkg->{version}, build => $pkg->{numbuild}, update => $version };
            }
        } elsif ($filter eq 'BUILD') {
	    if ($build_needed) {
                push @updates, { name => $pkg->{name}, installed => $pkg->{version}, build => $pkg->{numbuild}, update => $version, bump => $bump };
            }
        } else {
            if ($version_needed or $build_needed) {
                push @updates, { name => $pkg->{name}, installed => $pkg->{version}, build => $pkg->{numbuild}, update => $version, bump => $bump };
	    }
        }
    }
    return \@updates;
}

=head2 get_inst_names

  my @names = get_inst_names(get_available_updates());

C<get_inst_names()> returns a list of package names from an array reference, such
as one returned by C<get_available_updates()>.

=cut

# for a ref to an array of hashes of installed packages, return an array ref
# consisting of just their names
sub get_inst_names {
    script_error('get_inst_names requires an argument.') unless @_ == 1;
    my $inst = shift;
    my @installed;
    push @installed, $$_{name} for @$inst;
    return \@installed;
}

=head2 get_installed_cpans

  my @cpans = @{ get_installed_cpans() };

C<get_installed_cpans()> returns an array reference to a list of Perl
modules installed from the CPAN. This is used in C<sboinstall(1)> to
prevent conflicting installations from the CPAN and SlackBuilds.

=cut

# return a list of perl modules installed via the CPAN
sub get_installed_cpans {
  my @contents;
  for my $file (grep { -f $_ } map { "$_/perllocal.pod" } @INC) {
    my ($fh, $exit) = open_read($file);
    next if $exit;
    push @contents, grep {/Module/} <$fh>;
    close $fh;
  }
  my $mod_regex = qr/C<Module>\s+L<([^\|]+)/;
  my (@mods, @vers);
  for my $line (@contents) {
    push @mods, ($line =~ $mod_regex)[0];
  }
  return \@mods;
}

=head2 get_installed_packages

  my @packages = @{ get_installed_packages($type) };

C<get_installed_packages()> returns an array reference to a list of installed packages
matching the specified C<$type>. The available types are C<STD> for non-SBo packages,
C<SBO> for in-tree _SBo packages, C<DIRTY> for out-of-tree _SBo packages and C<ALL> for all.

The returned array reference holds a list of hash references representing
the names, versions and full installed package names of the returned packages.

=cut

# pull an array of hashes, each hash containing the name and version of a
# package currently installed. Gets filtered using STD, SBO, DIRTY or ALL.
sub get_installed_packages {
  script_error('get_installed_packages requires an argument.') unless @_ == 1;
  my $filter = shift;

  # Valid types: STD, SBO
  my (@pkgs, %types);
  foreach my $pkg (glob("$pkg_db/*")) {
    $pkg =~ s!^\Q$pkg_db/\E!!;
    my ($name, $version, $build) = $pkg =~ m#^([^/]+)-([^-]+)-[^-]+-([^-]+)$#
      or next;
    my $numbuild = $build;
    $numbuild =~ s/_SBo(|compat32)$//g ;
    push @pkgs, { name => $name, version => $version, build => $build, numbuild => $numbuild, pkg => $pkg };
    $types{$name} = 'STD';
  }

  # If we want all packages, let's just return them all
  return [ map { +{ name => $_->{name}, version => $_->{version}, build=> $_->{build}, numbuild => $_->{numbuild}, pkg => $_->{pkg} } } @pkgs ]
    if $filter eq 'ALL';

  # Otherwise, SlackBuilds with locations can be marked with SBO, and packages with
  # the _SBo tag but no location can be marked with DIRTY
  my @sbos = map { $_->{name} } grep { $_->{build} =~ m/_SBo(|compat32)$/ }
    @pkgs;
  if (@sbos) {
    my %locations = get_sbo_locations(map { s/-compat32//gr } @sbos);
    foreach my $sbo (@sbos) {
      $types{$sbo} = 'DIRTY';
      if ($locations{ $sbo =~ s/-compat32//gr }) {
         $types{$sbo} = 'SBO';
      }
    }
  }
  return [ map { +{ name => $_->{name}, version => $_->{version}, build => $_->{build}, numbuild => $_->{numbuild}, pkg => $_->{pkg} } }
    grep { $types{$_->{name}} eq $filter } @pkgs ];
}

=head2 get_local_outdated_versions

  my @outdated = get_local_outdated_versions($filter);

C<get_local_outdated_versions()> checks installed SBo packages from C<LOCAL_OVERRIDES>.
It returns an array with information about those that have version or build numbers
differing from the local repository or the SlackBuild in C<LOCAL_OVERRIDES>.

Build number differences with the SBo repository are returned, but are currently unused.
This subroutine is used only by C<sbocheck(1)>.

=cut

sub get_local_outdated_versions {
  script_error('get_local_outdated_versions requires an argument.') unless @_ == 1;
  my $filter = shift;
  my @outdated;

  my $local = $config{LOCAL_OVERRIDES};
  unless ( $local eq 'FALSE' ) {
    my $pkglist = get_installed_packages('SBO');
    my @local = grep { is_local($_->{name}) } @$pkglist;

    foreach my $sbo (@local) {
      my $local_location = get_sbo_location($sbo->{name});
      next if not defined $local_location;
      my $orig = get_orig_version($sbo->{name});
      next if not defined $orig;
      my $orig_build_number = get_orig_build_number($sbo->{name});
      my $local_build_number = get_sbo_build_number($local_location);

      if ($filter eq 'VERS') {
        next if not version_cmp($orig, $sbo->{version});
      } elsif ($filter eq 'BUILD' and defined $local_build_number) {
        if (not build_cmp($local_build_number, $sbo->{numbuild}, $orig, $sbo->{version})) {
	  next;
        }
      } elsif ($filter eq 'BOTH' and defined $local_build_number) {
	if (not build_cmp($local_build_number, $sbo->{numbuild}, $orig, $sbo->{version}) && not version_cmp($orig, $sbo->{version})) {
          next;
        }
      } else { next; }

      if (defined $orig_build_number and defined $local_build_number) {
        push @outdated, { %$sbo, orig => $orig, intree => $orig_build_number, bump => $local_build_number };
      } else {
        push @outdated, { %$sbo, orig => $orig };
      }
    }
  }

  return @outdated;
}

=head2 get_removed_builds

  my @removed = get_removed_builds();

C<get_removed_builds()> returns an array of SlackBuild names and versions of all out-of-tree
installed packages marked C<_SBo>.

=cut

# For each installed SlackBuild, find out whether it still exists in the tree
sub get_removed_builds {
    my @removed;
    my $pkg_list = get_installed_packages('DIRTY');

    for my $pkg (@$pkg_list) {
        push @removed, { name => $pkg->{name}, installed => $pkg->{version} };
    }

    return \@removed;
}

=head1 EXIT CODES

Pkgs.pm subroutines can return the following exit code:

  _ERR_SCRIPT        2   script or module bug

=head1 SEE ALSO

SBO::Lib(3), SBO::Lib::Build(3), SBO::Lib::Download(3), SBO::Lib::Info(3), SBO::Lib::Readme(3), SBO::Lib::Repo(3), SBO::Lib::Tree(3), SBO::Lib::Util(3)

=head1 AUTHORS

SBO::Lib was originally written by Jacob Pipkin <j@dawnrazor.net> with
contributions from Luke Williams <xocel@iquidus.org> and Andreas
Guldstrand <andreas.guldstrand@gmail.com>.

=head1 MAINTAINER

SBO::Lib is maintained by K. Eugene Carlson <kvngncrlsn@gmail.com>.

=head1 LICENSE

The sbotools are licensed under the MIT License.

Copyright (C) 2012-2017, Jacob Pipkin, Luke Williams, Andreas Guldstrand.

Copyright (C) 2024-2025, K. Eugene Carlson.

=cut

1;
